[Elevation profile] Add ability to render point cloud data as continuous elevation line

ViperMiniQ opened the pull request:
This PR adds the ability for the elevation profile of a point cloud layer to display a continuous elevation line (like vector or raster layers) instead of individual points.

Z values for the elevation line are derived from triangles - Delaunay triangulation is performed on the points within the selection zone (selection curve + tolerance buffer), and Z values are interpolated from the triangles that intersect the midpoint of the selection curve.

An option to extract the height from the triangulated surface has been added to the Elevation section of the layer properties. Here, users can choose to extract elevation data either directly from individual point cloud points or from the triangulated surface. 
The quality of the elevation line depends on the user’s selection: if a small number of points is included (for example, when a low tolerance setting is used), the resulting triangles may be of low quality and thus the elevation line won’t be as good. Users can adjust the tolerance setting to improve results. Triangulating the whole layer, especially considering the size of some datasets, can be too much, so this was chosen for the method.

The file delaunator.hpp had to be modified to make all functions inline - otherwise, the linker would complain. Is modifying 3rd party library acceptable, or would it be better to create a wrapper instead?

When experimenting with the elevation profile, you can configure a layer to use all or a subset of points within the selection zone (adjusted in symbology). 

You can also duplicate the layer, apply different symbology to it, and set it to render a subset of points. For example, it’s possible to use the same dataset twice in the elevation profile—one layer displaying all attributes and another showing only ground points. Doing that, you end up with something that looks like this:

<img width="3839" height="1779" alt="Screenshot_20260204_123700" src="https://github.com/user-attachments/assets/25f01435-bb62-4f24-b192-e7932f5656f4" />


<img width="1319" height="603" alt="Screenshot_20260204_123931" src="https://github.com/user-attachments/assets/9b6a2648-1944-4c10-b429-5100bcd756d5" />



Edit: fixes #64761

ViperMiniQ committed:
set delaunator function as inline

this prevents linker issues when including the delaunator in other files which have includes that also include delaunator

ViperMiniQ committed:
add pointcloud profile type enum

ViperMiniQ committed:
add type to profile generators

ViperMiniQ committed:
add triangulated point cloud surface elevation extraction generator

this generator triangulates the gathered points from which the height is extracted

interpolateZ function is added to get the height at point within a triangle

ViperMiniQ committed:
add and connect GUI elements for triangulated point cloud elevation profile generation

ViperMiniQ committed:
document functions, revert accidental changes to individual point cloud elevation generator

ViperMiniQ committed:
fix string, use base identify

ViperMiniQ committed:
fix docs, fix strings

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.h:
```diff
@@ -207,17 +211,174 @@ class CORE_EXPORT QgsPointCloudLayerElevationProperties : public QgsMapLayerElev
+     * \see renderType()
+     * \since QGIS 4.0
+     */
+    void setRenderType( Qgis::PointCloudProfileType type ) { mRenderType = type; }
```
For consistency, I would suggest to rename to setType() + type() + mType as `QgsVectorLayerElevationProperties` uses that naming for the same thing...

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.h:
```diff
@@ -207,17 +211,174 @@ class CORE_EXPORT QgsPointCloudLayerElevationProperties : public QgsMapLayerElev
+
+    /**
+     * Returns the symbol used to render lines for the layer in elevation profile plots.
+     *
```
worth adding here a note this is only valid for triangulated surface + when profileSymbology() is Line.

Similarly for all the other added getters/setters that depend on triangulated surface and type of profileSymbology()

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.cpp:
```diff
@@ -105,8 +260,13 @@ QgsPointCloudLayerElevationProperties *QgsPointCloudLayerElevationProperties::cl
   res->mPointSymbol = mPointSymbol;
   res->mPointColor = mPointColor;
   res->mRespectLayerColors = mRespectLayerColors;
+  res->mRenderType = mRenderType;
```
it seems some member variables are not being copied (e.g. mSymbology, mElevationLimit)

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.cpp:
```diff
@@ -58,6 +172,22 @@ QDomElement QgsPointCloudLayerElevationProperties::writeXml( QDomElement &parent
   element.setAttribute( u"point_color"_s, QgsColorUtils::colorToString( mPointColor ) );
   element.setAttribute( u"respect_layer_colors"_s, mRespectLayerColors ? u"1"_s : u"0"_s );
   element.setAttribute( u"opacity_by_distance"_s, mApplyOpacityByDistanceEffect ? u"1"_s : u"0"_s );
+  element.setAttribute( u"render_type"_s, qgsEnumValueToKey( mRenderType ) );
```
it seems mSymbology is not being written to and read from xml

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -794,4 +799,574 @@ void QgsPointCloudLayerProfileGenerator::visitBlock( const QgsPointCloudBlock *b
+  for ( QgsPoint point : mGatheredPoints )
+  {
+    vertices.push_back( point.x() );
+    vertices.push_back( -point.y() ); // flipping y to have correctly oriented triangles from delaunator
```
flipping y here, but then we're flipping y again in the for loop below? can't we skip both?

wonder-sk reviewed (commented):
There are some minor GUI issues:
- initially when the elevation properties get open with traingulated surface:
  - when using line symbology -> GUI also shows fill marker, limit
  - when using fill below symbology -> still shows line marker
- the limit - should say "not set" instead of 0 (like for vector layers GUI)

ViperMiniQ committed:
remove marker symbol, rename renderType to type

ViperMiniQ committed:
create base pointcloudprofilegenerator

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.h:
```diff
@@ -207,17 +211,133 @@ class CORE_EXPORT QgsPointCloudLayerElevationProperties : public QgsMapLayerElev
+     * \see setRenderType()
+     * \since QGIS 4.0
+     */
+    Qgis::PointCloudProfileType renderType() const { return mType; }
```
let's rename to `type()` for consistency

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerelevationproperties.h:
```diff
@@ -207,17 +211,133 @@ class CORE_EXPORT QgsPointCloudLayerElevationProperties : public QgsMapLayerElev
+    /**
+     * Returns the profile type used when generating elevation profile plots.
+     *
+     * \see setRenderType()
```
let's fix the docstring references (also for setType())

ViperMiniQ committed:
address review

- remove marker symbol and related getters and setters
- rename renderType to type
- fix docs
- pure virtual visitblocks
- remove unused imports

ViperMiniQ committed:
add test

ViperMiniQ committed:
override constructor

ViperMiniQ committed:
remove duplicate attribute, fix docs

ViperMiniQ committed:
clang tidy, minor fixes

ViperMiniQ committed:
reset results only if canceled or no points gathered

- fixes a bug where if we failed to get a single point from any single file of a vpc, we would return early and scrap the results from all other files

- fixes test accordingly

ViperMiniQ committed:
add docs to base class

ViperMiniQ committed:
fix docs param naming

ViperMiniQ committed:
spelling

ptitjano commented:
@ViperMiniQ Thank you for working on this. Could` you please add a unit test to `QgsGeometryUtils::interpolateZ`?
